### PR TITLE
Fixes problems with multiple sources of alpha on humans

### LINKED ISSUE #17915: Vampires are unable to use Chameleon Skin gene
**Issue Description**:
Vampires do not become invisible when using Chameleon Skin. (No vampire class has been picked when I encountered the bug).

**What did you expect to happen**: 
For the vampire to become invisible.

**What happened instead**:
Gene had no effect beyond the "You feel like you're fading into your surroundings" message. 

**Why is this bad/What are the consequences**:
Frankly speaking, it's not a common issue, but it's likely something that will make a vampire geneticist sad.

**Steps to reproduce the problem**:
Be a vampire. Get the gene. See yourself still not being invisible.

**When did the problem start happening**:
Presumably it's a leftover from the old vampires, who had Darkness Cloak (and which is incompatible with Chameleon), which is now restricted to Umbrae. 

**Extra information**:

### LINKED ISSUE #26422: Changelings chameleon skin cancels Environmental Adaptation
### Exploit Reports

- [X] I confirm this issue is not an exploit. (Required)

### BYOND Version

515.1640

### Issue Description

Changelings chameleon skin and Environmental Adaptation work very weird together, like some times they can stop even working, and you will need to tern them off and on

### What did you expect to happen?

they both should work correctly together

### What happened instead?

canceling chameleon skin cancels Environmental Adaptation 

### Why is this bad/What are the consequences?

_No response_

### Steps to reproduce the issue.

be Changeling -> take chameleon skin and Environmental Adaptation -> use them together -> cancel chameleon skin 

### When did the problem start happening?

¯\_(ツ)_/¯

### Extra information

_No response_

### Relevant log output/runtime error

_No response_

### PR DESCRIPTION
**Pull request by Contrabang**
## What Does This PR Do

Technical alternate solution to #25882
Fixes #26422. Fixes #17915.

This PR makes it so that alpha is tracked from each source, and your alpha is set to the minimum. This includes
- MOD stealth module
- RD's stealth armor
- Dantilion's Deploy Decoy
- Umbrae's Cloak of Darkness
- Heirophant staff teleporting
- Potted plants
- Chameleon Gene
- Shadow Gene

## Why It's Good For The Game

"Changeling abilities should not fuck with each other. Other stealth abilities should not unstealth you, requiring to turn your ability on and off again. A fully stealth build should mesh together better.

This also means you can wear reactive armor without needing to turn the ability on or off, wear a modsuit stealth module, use a hiero staff, etc etc." - #25882

## Testing

Put on most of these things at the same time, and I always had the lowest alpha. However, I only went completely invisible with the Chameleon and Shadow genes were at 0.

## Changelog
:cl:
fix: Picking up a plant while using Umbrae's Cloak of Darkness will no longer show you behind the plant
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/human/human_defines.dm [hunk 1]
```diff
@@ -103,3 +103,5 @@
 	var/list/bodyparts = list()
 	/// map organ names to organs
 	var/list/bodyparts_by_name = list()
+
+	var/list/alpha_sources
```
Comment (Sirryan2002) [inline code/modules/mob/living/carbon/human/human_defines.dm:107]: document variable, please also indicate that it's a lazylist !

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 2]
```diff
@@ -2059,3 +2059,30 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
...
+		alpha = get_alpha()
+
+/**
+ * Gets the target alpha of the human
+ *
+ * optional_key - use this to get the alpha of an exact source
+ */
+/mob/living/carbon/human/proc/get_alpha(optional_key)
+	if(optional_key)
+		return alpha_sources[optional_key]
```
Comment (Sirryan2002) [inline code/modules/mob/living/carbon/human/human_mob.dm:2085]: is there a chance that `optional_key` is not stored in alpha_sources when this proc is called? Or is this meant to be an unsafe proc that runtimes if you provided an optional key that's not on the mob?

**Review by Sirryan2002 — APPROVED** [review]

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 3]
```diff
@@ -2059,3 +2059,30 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
 	dna.ResetUIFrom(src)
 	flavor_text = ""
 
+/**
+ * Helper for tracking alpha, use this to set an alpha source
+ *
+ * alpha_num - num between 0 and 255 that represents the alpha from `source`
+ * source - a typepath or other key that is the alpha source. Maybe this should use UID someday
```
Comment (hal9000PR) [inline code/modules/mob/living/carbon/human/human_mob.dm:2066]: I was just about to ask why isnt this a UID

**Review by DGamerL — CHANGES_REQUESTED** [review]
Please respond to reviews

**Comment by Contrabang** [comment]
I don't have the motivation or time to maintain any of my PRs

**Comment by Contrabang** [comment]
I'm reviving this, fuck it

**Comment by Adrer** [comment]
Should fix #17915

File: code/game/dna/mutations/mutation_powers.dm [hunk 4]
```diff
@@ -209,19 +209,26 @@
 	..()
 	block = GLOB.shadowblock
 
+/datum/mutation/stealth/darkcloak/deactivate(mob/living/M)
+	..()
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		H.set_alpha_tracking(ALPHA_VISIBLE, src)
+
```
Comment (BiancaWilkson) [inline code/game/dna/mutations/mutation_powers.dm:217]:
```suggestion
	if(!ishuman(M))
		return
	var/mob/living/carbon/human/H = M
	H.set_alpha_tracking(ALPHA_VISIBLE, src)
```

File: code/game/dna/mutations/mutation_powers.dm [hunk 5]
```diff
@@ -209,19 +209,26 @@
...
+			H.set_alpha_tracking(H.get_alpha() * 0.8, src)
 	else
-		M.reset_visibility()
-		M.alpha = round(255 * 0.8)
-	if(M.alpha == 0)
-		M.make_invisible()
+		H.reset_visibility()
+		H.set_alpha_tracking(ALPHA_VISIBLE * 0.8, src)
+	if(H.get_alpha(src) == 0)
+		H.make_invisible()
```
Comment (BiancaWilkson) [inline code/game/dna/mutations/mutation_powers.dm:231]: make_invisible directly sets the mob's alpha to 128, will that cause problems for this system?

**Review by BiancaWilkson — COMMENTED** [review]
A small guard clause change, and want to make sure of something I noticed. Otherwise code looks good

File: code/game/dna/mutations/mutation_powers.dm (hunk 5 above)
Comment (Contrabang) [reply code/game/dna/mutations/mutation_powers.dm:231]: No, that alpha will be reset when they become non-invisible

**Review by BiancaWilkson — APPROVED** [review]

**Review by DGamerL — APPROVED** [review]

**Pull request merged**